File: apps/ios/RelayKit/StoragePolicy.swift
import Foundation

/// Storage policy for the glasses' 4 GB.
///
/// `docs/APPS-SCOPE.md` §3.2, which is short enough to quote: poll disk state
/// (`0x0909` / `0x091C`), reserve headroom for audio and warn before video eats
/// it, sync-and-free proactively rather than waiting to hit full, and **never
/// delete un-uploaded audio** — `0x0911` 清除未上传文件 exists precisely because
/// the firmware tracks that distinction.
///
/// The asymmetry that drives all of it: 1080p video runs ~4.5 GB/h and Opus
/// audio runs ~10.8 MB/h, so **video is about four hundred times more
/// expensive**. Forty minutes of video is a whole day of conversation. That is
/// not a warning to bury in a settings screen.

public enum CaptureFormat: String, Sendable, Equatable {
    /// ~24 kbps mono. `docs/APPS-SCOPE.md` §3.1.
    case opus
    /// 16 kHz / 16-bit mono.
    case pcm16

    /// Bytes per second on the glasses' own storage.
    public var bytesPerSecond: Int {
        switch self {
        case .opus: return 3_000
        case .pcm16: return 32_000
        }
    }
}

/// ~4.5 GB/h at 1080p.
public let videoBytesPerSecond = 1_342_177

/// The capture-worthy part of a day, per `docs/APPS-SCOPE.md` §3.1.
public let captureDaySeconds = 16 * 3_600

public struct StorageBudget: Sendable, Equatable {
    /// What the glasses record audio as. Still an open question against real
    /// hardware (§8), and it changes every number here by ~10×, so it is a
    /// parameter rather than a constant.
    public var format: CaptureFormat
    /// Days of audio that video is not allowed to take. One by default: the
    /// product promise is "records your day", so the day is the floor.
    public var reservedDays: Double
    /// Below this fraction of the reserve, stop being polite about it.
    public var criticalFraction: Double

    public init(format: CaptureFormat = .opus, reservedDays: Double = 1, criticalFraction: Double = 0.25) {
        self.format = format
        self.reservedDays = reservedDays
        self.criticalFraction = criticalFraction
    }

    /// Bytes that must stay available for audio.
    public var reservedBytes: Int {
        Int(Double(captureDaySeconds * format.bytesPerSecond) * reservedDays)
    }
}

public enum StorageLevel: String, Sendable, Equatable, Comparable {
    case ok
    case watch
    /// Below the audio reserve. Video is refused from here down.
    case warn
    /// Deep below it. Sync now, and say so out loud.
    case critical

    private var rank: Int {
        switch self {
        case .ok: return 0
        case .watch: return 1
        case .warn: return 2
        case .critical: return 3
        }
    }

    public static func < (lhs: StorageLevel, rhs: StorageLevel) -> Bool {
        lhs.rank < rhs.rank
    }
}

public struct StorageAssessment: Sendable, Equatable {
    public let level: StorageLevel
    public let freeBytes: Int
    public let totalBytes: Int
    /// How much more audio fits, in seconds.
    public let audioSecondsRemaining: Int
    /// How much video fits before the audio reserve is breached. Zero once it
    /// already is.
    public let videoSecondsBeforeReserve: Int
    /// False once video would come out of the audio reserve.
    public let videoAllowed: Bool
    /// Bytes held by files the device says are already uploaded. This is the
    /// only space the app may reclaim on its own.
    public let reclaimableBytes: Int
    /// Bytes that exist only on the glasses. Never touched.
    public let unsyncedBytes: Int
    public let message: String
}

public enum StoragePolicy {

    public static func assess(
        disk: DiskInfo,
        files: [RemoteFile] = [],
        budget: StorageBudget = StorageBudget()
    ) -> StorageAssessment {
        let reserve = budget.reservedBytes
        let audioRate = budget.format.bytesPerSecond
        let audioSeconds = disk.freeBytes / max(1, audioRate)
        let spare = max(0, disk.freeBytes - reserve)
        let videoSeconds = spare / videoBytesPerSecond

        let reclaimable = files.filter(\.uploaded).reduce(0) { $0 + $1.sizeBytes }
        let unsynced = files.filter { !$0.uploaded }.reduce(0) { $0 + $1.sizeBytes }

        let level: StorageLevel
        if disk.freeBytes < Int(Double(reserve) * budget.criticalFraction) {
            level = .critical
        } else if disk.freeBytes < reserve {
            level = .warn
        } else if disk.freeBytes < reserve * 2 {
            level = .watch
        } else {
            level = .ok
        }

        return StorageAssessment(
            level: level,
            freeBytes: disk.freeBytes,
            totalBytes: disk.totalBytes,
            audioSecondsRemaining: audioSeconds,
            videoSecondsBeforeReserve: videoSeconds,
            videoAllowed: level < .warn && videoSeconds > 0,
            reclaimableBytes: reclaimable,
            unsyncedBytes: unsynced,
            message: message(level: level, audioSeconds: audioSeconds, videoSeconds: videoSeconds)
        )
    }

    private static func message(level: StorageLevel, audioSeconds: Int, videoSeconds: Int) -> String {
        let audioHours = Double(audioSeconds) / 3_600
        switch level {
        case .ok:
            return String(format: "About %.0f hours of audio left on the glasses.", audioHours)
        case .watch:
            return String(
                format: "About %.0f hours of audio left. Video would use it up in %d minutes.",
                audioHours,
                max(0, videoSeconds / 60)
            )
        case .warn:
            return String(
                format: "Under a day of audio left (%.0f hours). Video is off until the "
                    + "glasses sync — a few minutes of it would cost you the day.",
                audioHours
            )
        case .critical:
            return String(
                format: "The glasses are nearly full — about %.0f hours left. Plug them in "
                    + "so tonight's sync can free the space.",
                audioHours
            )
        }
    }

    /// The only files the app may delete, ever.
    ///
    /// `RemoteFile.uploaded` is the *device's* opinion, not ours. The firmware
    /// tracks it and `0x0911` acts on it, so deleting on the strength of our own
    /// bookkeeping is how the only copy of an afternoon disappears — our
    /// bookkeeping is exactly what a crash loses.
    public static func safeToDelete(_ files: [RemoteFile]) -> [RemoteFile] {
        files.filter(\.uploaded)
    }

    /// Oldest-uploaded-first, until `targetBytes` is met. Returns fewer than
    /// asked for — possibly none — rather than reaching past the uploaded set.
    ///
    /// `listFiles` returns device order, which is creation order on this
    /// firmware. That is the only ordering available: `RemoteFile` carries no
    /// timestamp, because the reply layout for the file-list command is one of
    /// the payloads `docs/APPS-SCOPE.md` §5.1 marks as unattested.
    public static func reclaim(targetBytes: Int, from files: [RemoteFile]) -> [RemoteFile] {
        var chosen: [RemoteFile] = []
        var freed = 0
        for file in safeToDelete(files) {
            if freed >= targetBytes { break }
            chosen.append(file)
            freed += file.sizeBytes
        }
        return chosen
    }

    /// Would recording video for this long push audio below its reserve?
    ///
    /// The question the UI has to answer *before* the shutter, not after.
    public static func videoWouldBreachReserve(
        seconds: Int,
        disk: DiskInfo,
        budget: StorageBudget = StorageBudget()
    ) -> Bool {
        let cost = seconds * videoBytesPerSecond
        return disk.freeBytes - cost < budget.reservedBytes
    }
}

// MARK: - polling

/// Polls `getDiskInfo` and reports assessments.
///
/// A poll rather than a subscription because the device only reports disk state
/// when asked (`0x0909` / `0x091C` are commands, not notifications), and a
/// product that discovers a full device at the end of the day has already lost
/// the afternoon.
///
/// Five minutes by default: at 4.5 GB/h a video session can consume the whole
/// reserve inside one interval, so the recording paths ask directly rather than
/// waiting for the next tick.
public actor StorageMonitor {

    private let glasses: GlassesTransport
    private let clock: RelayClock
    private let intervalMs: Int
    private let budget: StorageBudget
    private let onAssessment: @Sendable (StorageAssessment) -> Void

    private var running = false
    private var lastLevel: StorageLevel?
    public private(set) var latest: StorageAssessment?

    public init(
        glasses: GlassesTransport,
        clock: RelayClock = SystemClock(),
        intervalMs: Int = 5 * 60 * 1_000,
        budget: StorageBudget = StorageBudget(),
        onAssessment: @escaping @Sendable (StorageAssessment) -> Void = { _ in }
    ) {
        self.glasses = glasses
        self.clock = clock
        self.intervalMs = intervalMs
        self.budget = budget
        self.onAssessment = onAssessment
    }

    public var isRunning: Bool { running }

    /// Whether the level changed on the last sample. The UI nags on a change,
    /// not on every poll — a warning repeated every five minutes is a warning
    /// nobody reads.
    public private(set) var levelChanged = false

    public func start() {
        guard !running else { return }
        running = true
        Task { [weak self] in
            await self?.loop()
        }
    }

    public func stop() {
        running = false
    }

    /// One sample, on demand. Called before starting video, and after a sync.
    @discardableResult
    public func sample() async -> StorageAssessment? {
        guard let disk = try? await glasses.getDiskInfo() else { return nil }
        let files = (try? await glasses.listFiles()) ?? []
        let assessment = StoragePolicy.assess(disk: disk, files: files, budget: budget)
        levelChanged = lastLevel != assessment.level
        lastLevel = assessment.level
        latest = assessment
        onAssessment(assessment)
        return assessment
    }

    private func loop() async {
        while running {
            await sample()
            await clock.sleep(ms: intervalMs)
        }
    }
}

File: apps/android/relay-bridge/src/main/java/glass/relay/bridge/storage/StoragePolicy.kt
package glass.relay.bridge.storage

import kotlin.math.max
import kotlin.math.min

/**
 * Who gets the glasses' 4 GB, and what happens as it runs out.
 *
 * `APPS-SCOPE.md` §3.2 makes this a requirement rather than a nicety, and the
 * arithmetic is why. The device holds 4 GB, shared between audio, photos and
 * video:
 *
 * | | Rate | 4 GB holds |
 * |---|---|---|
 * | Opus audio ~24 kbps | ~10.8 MB/h | ~370 h |
 * | PCM 16 kHz/16-bit | ~115 MB/h | ~35 h |
 * | 1080p video | **~4.5 GB/h** | ~53 min |
 *
 * So **fourteen minutes of video costs a full day of Opus audio**, and video is
 * the only thing on the device that can consume storage faster than a person
 * can notice. That single fact drives every rule below.
 *
 * ## The rules
 *
 *  1. **Reserve headroom for audio.** [Config.reserveAudioHours] of recording
 *     is set aside; video is measured against what is left over, not against
 *     free space.
 *  2. **Warn before video eats it**, with a number: "this will leave you 3
 *     hours of audio". A percentage bar tells nobody anything useful.
 *  3. **Sync and free proactively**, at [Config.syncAtFreeFraction], not when
 *     the device is full. A sync started at 100% full has nowhere to put the
 *     recording that is still running.
 *  4. **Never propose deleting un-uploaded audio.** `0x0911` 清除未上传文件
 *     exists precisely because the firmware tracks that distinction, and the
 *     existence of that command is not permission to use it. `StoragePolicyTest`
 *     asserts that no assessment, at any level of pressure, ever suggests it.
 *
 * Everything here is arithmetic on a snapshot, so it is testable without a
 * device — which matters, because the interesting cases are the ones you cannot
 * reach on a bench without filling a real 4 GB first.
 */
object StoragePolicy {

    /** Bytes per second, from `SYSTEM.md` §3.1 and `APPS-SCOPE.md` §3.1. */
    object Rates {
        /** Opus ~24 kbps mono, the format the SDK's encoders suggest. */
        const val AUDIO_OPUS: Long = 3_000

        /** PCM 16 kHz 16-bit mono — the pessimistic bound. */
        const val AUDIO_PCM16: Long = 32_000

        /** 1080p, ~4.5 GB/h. The one that ruins a day. */
        const val VIDEO_1080P: Long = 1_342_177

        /** A full-resolution stills capture. Rough, and only used for warnings. */
        const val PHOTO_BYTES: Long = 3L * 1024 * 1024
    }

    const val DEVICE_TOTAL_BYTES: Long = 4L * 1024 * 1024 * 1024

    enum class RecordingFormat(val bytesPerSecond: Long) {
        Opus(Rates.AUDIO_OPUS),
        Pcm16(Rates.AUDIO_PCM16),
    }

    data class Config(
        /**
         * Hours of audio to keep room for at all times.
         *
         * A working day plus slack. In Opus this is ~130 MB of a 4 GB device —
         * cheap insurance against a video session that runs long.
         */
        val reserveAudioHours: Double = 12.0,

        /** Start the sync ritual once free space drops below this fraction. */
        val syncAtFreeFraction: Double = 0.25,

        /**
         * Critical once free space falls below this fraction of the audio
         * reserve.
         *
         * Expressed against the reserve rather than against total capacity on
         * purpose. A flat "8% of the disk" is 327 MB on this device, which is
         * two and a half times the twelve-hour reserve — so a percentage
         * threshold would declare a critical state while there was still half a
         * week of audio headroom, and `Low` would be unreachable. Every level
         * here is measured in hours of audio, because that is the unit the
         * product is denominated in.
         */
        val criticalReserveFraction: Double = 0.25,

        val format: RecordingFormat = RecordingFormat.Opus,
    )

    /** What `0x0909` / `0x091C` reports. */
    data class DiskSnapshot(val totalBytes: Long, val freeBytes: Long) {
        val usedBytes: Long get() = max(0, totalBytes - freeBytes)
        val freeFraction: Double get() = if (totalBytes <= 0) 0.0 else freeBytes.toDouble() / totalBytes
    }

    /**
     * What is on the device, from `0x0E01` 获取文件列表.
     *
     * `uploaded` is the *firmware's* answer, not ours. Trusting our own
     * bookkeeping here is how the only copy of something disappears.
     */
    data class Inventory(
        val unuploadedAudioBytes: Long = 0,
        val uploadedAudioBytes: Long = 0,
        val videoBytes: Long = 0,
        val photoBytes: Long = 0,
    ) {
        val reclaimableBytes: Long get() = uploadedAudioBytes
    }

    enum class Level {
        /** Room for the reserve and then some. */
        Healthy,

        /** Reserve intact, but the sync threshold has been crossed. */
        Watch,

        /** The audio reserve is being eaten. */
        Low,

        /** No room for a working day of audio. */
        Critical,
    }

    sealed interface Action {
        /** Run the two-phase WiFi sync now, while it can still finish. */
        data class SyncNow(val why: String) : Action

        /** Delete audio the box has confirmed it holds. Never the rest. */
        data class FreeUploadedAudio(val bytes: Long) : Action

        /** Video is allowed but will cost this much of the audio reserve. */
        data class WarnBeforeVideo(val minutesAvailable: Int, val audioHoursAfter: Double) : Action

        /** Video would break the audio reserve. Refuse it. */
        data class BlockVideo(val why: String) : Action
    }

    data class Assessment(
        val level: Level,
        val freeBytes: Long,
        /** Hours of audio the remaining space holds at the current format. */
        val audioHoursRemaining: Double,
        /** Minutes of 1080p before the audio reserve is broken. Often small. */
        val videoMinutesBeforeReserveLost: Int,
        val actions: List<Action>,
        val warnings: List<String>,
    ) {
        val videoAllowed: Boolean get() = actions.none { it is Action.BlockVideo }
    }

    fun assess(
        disk: DiskSnapshot,
        inventory: Inventory = Inventory(),
        config: Config = Config(),
    ): Assessment {
        val audioRate = config.format.bytesPerSecond
        val reserveBytes = (config.reserveAudioHours * 3600 * audioRate).toLong()
        val free = max(0, disk.freeBytes)

        val audioHoursRemaining = free.toDouble() / audioRate / 3600.0
        val spareOverReserve = free - reserveBytes
        val videoSecondsBeforeReserveLost = max(0, spareOverReserve) / Rates.VIDEO_1080P
        val videoMinutes = (videoSecondsBeforeReserveLost / 60).toInt()

        val level = when {
            free < reserveBytes * config.criticalReserveFraction -> Level.Critical
            free < reserveBytes -> Level.Low
            free < reserveBytes * 3 || disk.freeFraction <= config.syncAtFreeFraction -> Level.Watch
            else -> Level.Healthy
        }

        val actions = mutableListOf<Action>()
        val warnings = mutableListOf<String>()

        if (level != Level.Healthy) {
            actions += Action.SyncNow(
                when (level) {
                    Level.Critical -> "the glasses are nearly full — sync now or capture will stop"
                    Level.Low -> "free space has dropped into the audio reserve"
                    else -> "free space is below ${(config.syncAtFreeFraction * 100).toInt()}%"
                },
            )
        }

        // The only deletion this policy will ever propose. Audio the box has
        // confirmed it holds is a copy; everything else is an original.
        if (inventory.reclaimableBytes > 0 && level != Level.Healthy) {
            actions += Action.FreeUploadedAudio(inventory.reclaimableBytes)
        }

        if (inventory.unuploadedAudioBytes > 0 && level == Level.Critical) {
            warnings += "there is ${mb(inventory.unuploadedAudioBytes)} of audio the box has " +
                "never seen — it will not be deleted, so sync before recording video"
        }

        when {
            spareOverReserve <= 0 -> actions += Action.BlockVideo(
                "recording video would eat into the ${config.reserveAudioHours.toInt()} hours " +
                    "of audio headroom the day needs",
            )
            videoMinutes < 5 -> actions += Action.WarnBeforeVideo(
                minutesAvailable = videoMinutes,
                audioHoursAfter = audioHoursAfterVideo(free, videoMinutes, audioRate),
            )
            else -> {
                // Even with room, say what it costs: a few minutes of 1080p is
                // measured in hours of audio, and nobody guesses that correctly.
                actions += Action.WarnBeforeVideo(
                    minutesAvailable = videoMinutes,
                    audioHoursAfter = audioHoursAfterVideo(free, min(videoMinutes, 10), audioRate),
                )
            }
        }

        if (inventory.videoBytes > free) {
            warnings += "video already occupies ${mb(inventory.videoBytes)}, more than the " +
                "${mb(free)} still free"
        }

        return Assessment(
            level = level,
            freeBytes = free,
            audioHoursRemaining = audioHoursRemaining,
            videoMinutesBeforeReserveLost = videoMinutes,
            actions = actions,
            warnings = warnings,
        )
    }

    /**
     * How long a minute of 1080p costs in hours of audio.
     *
     * The number the warning copy needs. At 4.5 GB/h against Opus at 10.8 MB/h,
     * one minute of video is a little over seven hours of audio — which is the
     * whole argument for §3.2 in one figure.
     */
    fun audioHoursPerVideoMinute(format: RecordingFormat = RecordingFormat.Opus): Double =
        (Rates.VIDEO_1080P * 60.0) / format.bytesPerSecond / 3600.0

    private fun audioHoursAfterVideo(freeBytes: Long, videoMinutes: Int, audioRate: Long): Double {
        val afterVideo = max(0, freeBytes - videoMinutes * 60L * Rates.VIDEO_1080P)
        return afterVideo.toDouble() / audioRate / 3600.0
    }

    private fun mb(bytes: Long): String = "%.0f MB".format(bytes / 1024.0 / 1024.0)
}
